packages/beacon-node/src/chain/validation/executionPayloadEnvelope.ts

```diff
@@ -0,0 +1,120 @@
+  chain: IBeaconChain,
+  executionPayloadEnvelope: gloas.SignedExecutionPayloadEnvelope
+): Promise<void> {
+  return validateExecutionPayloadEnvelope(chain, executionPayloadEnvelope);
+}
+
+async function validateExecutionPayloadEnvelope(
+  chain: IBeaconChain,
+  executionPayloadEnvelope: gloas.SignedExecutionPayloadEnvelope
+): Promise<void> {
+  const envelope = executionPayloadEnvelope.message;
+  const {builderIndex, payload} = envelope;
+  const blockRootHex = toRootHex(envelope.beaconBlockRoot);
+
+  //  GLOAS: [IGNORE] The envelope's block root `envelope.block_root` has been seen (via
+  //  gossip or non-gossip sources) (a client MAY queue payload for processing once
+  //  the block is retrieved).
+  // TODO GLOAS: Need to review this
+  const block = chain.forkChoice.getBlock(envelope.beaconBlockRoot);
+  if (block === null) {
+    throw new ExecutionPayloadEnvelopeError(GossipAction.IGNORE, {
+      code: ExecutionPayloadEnvelopeErrorCode.BLOCK_ROOT_UNKNOWN,
+      blockRoot: blockRootHex,
+    });
+  }
+
+  //  [IGNORE] The node has not seen another valid
+  //  `SignedExecutionPayloadEnvelope` for this block root from this builder.
+  if (chain.seenExecutionPayloadEnvelopes.isKnown(blockRootHex, builderIndex)) {
+    throw new ExecutionPayloadEnvelopeError(GossipAction.IGNORE, {
+      code: ExecutionPayloadEnvelopeErrorCode.ENVELOPE_ALREADY_KNOWN,
+      blockRoot: blockRootHex,
+      slot: envelope.slot,
+      builderIndex,
+    });
+  }
+
+  //  [IGNORE] The envelope is from a slot greater than or equal to the latest finalized slot -- i.e. validate that `envelope.slot >= compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)`
+  const finalizedCheckpoint = chain.forkChoice.getFinalizedCheckpoint();
+  const finalizedSlot = computeStartSlotAtEpoch(finalizedCheckpoint.epoch);
+  if (envelope.slot < finalizedSlot) {
+    throw new ExecutionPayloadEnvelopeError(GossipAction.IGNORE, {
+      code: ExecutionPayloadEnvelopeErrorCode.BELONG_TO_FINALIZED_BLOCK,
+      envelopeSlot: envelope.slot,
+      finalizedSlot,
+    });
+  }
+
+  // [REJECT] `block` passes validation.
+  // TODO GLOAS: implement this. Technically if we cannot get proto block from fork choice,
+  // it is possible that the block didn't pass the validation
+
+  // [REJECT] `block.slot` equals `envelope.slot`.
+  if (block.slot !== envelope.slot) {
+    throw new ExecutionPayloadEnvelopeError(GossipAction.REJECT, {
+      code: ExecutionPayloadEnvelopeErrorCode.SLOT_MISMATCH,
+      envelopeSlot: envelope.slot,
+      blockSlot: block.slot,
+    });
+  }
+
+  if (block.builderIndex === undefined || block.blockHashHex === undefined) {
+    // Cache fail. This should not happen
+    throw new ExecutionPayloadEnvelopeError(GossipAction.IGNORE, {
+      code: ExecutionPayloadEnvelopeErrorCode.CACHE_FAIL,
+      blockRoot: blockRootHex,
+    });
+  }
+
+  // [REJECT] `envelope.builder_index == bid.builder_index`
+  if (builderIndex !== block.builderIndex) {
```

Comment:
more readable to have as per spec


```suggestion
  if (envelope.builderIndex !== block.builderIndex) {
```